src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -1361,29 +1362,43 @@ private void DecodeCustomAttributeParameterTypeOrThrow(ref BlobReader sigReader,
 
                 case SignatureTypeCode.TypeHandle:
                     // The type of the parameter can either be an enum type or System.Type.
-                    bool isNoPiaLocalType;
                     type = GetSymbolForTypeHandleOrThrow(sigReader.ReadTypeHandle(), out isNoPiaLocalType, allowTypeSpec: true, requireShortForm: true);
+                    break;
 
-                    var underlyingEnumType = GetEnumUnderlyingType(type);
-
-                    // Spec: If the parameter kind is an enum -- simply store the value of the enum's underlying integer type.
-                    if ((object)underlyingEnumType != null)
+                case SignatureTypeCode.GenericTypeParameter:
+                    if (!sigReader.TryReadCompressedInteger(out int paramPosition))
                     {
-                        Debug.Assert(!isNoPiaLocalType);
-
-                        // GetEnumUnderlyingType always returns a valid enum underlying type
-                        typeCode = underlyingEnumType.SpecialType.ToSerializationType();
-                        return;
+                        throw new UnsupportedSignatureContent();
                     }
 
-                    if ((object)type == SystemTypeSymbol)
-                    {
-                        Debug.Assert(!isNoPiaLocalType);
-                        typeCode = SerializationTypeCode.Type;
-                        return;
-                    }
+                    type = GetGenericTypeArgumentSymbol(paramPosition);
```

Comment:
This is going to be long.

First, while this appear to be doing the right thing, the fix is incomplete. For example, the following two tests are failing:
```
        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66370")]
        public void Attribute_GenericParameter_03()
        {
            var source = """
                public class A2<T> : System.Attribute
                {
                    public A2(T t) { }
                }

                public enum E { }
                
                [A2<object>(C1.C)]
                class C1
                {
                    public const E C = (E)33;
                }
                """;

            var verifier = CompileAndVerify(source, symbolValidator: static module =>
            {
                var c = module.GlobalNamespace.GetTypeMember("C1");
                var attr = c.GetAttributes().Single(d => d.AttributeClass?.Name == "A2");
                Assert.False(attr.HasErrors);
                //Assert.Empty(attr.NamedArguments);
                //var arg = attr.ConstructorArguments.Single();
                //Assert.Equal(33, arg.Value);
            });
            verifier.VerifyDiagnostics();
        }

        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66370")]
        public void Attribute_GenericParameter_04()
        {
            var source = """
                public class A2<T> : System.Attribute
                {
                    public A2(T t) { }
                }

                [A2<int>(33)]
                class C1
                {
                }
                """;

            var verifier = CompileAndVerify(source, symbolValidator: static module =>
            {
                var c = module.GlobalNamespace.GetTypeMember("C1");
                var attr = c.GetAttributes().Single(d => d.AttributeClass?.Name == "A2");
                Assert.False(attr.HasErrors);
                //Assert.Empty(attr.NamedArguments);
                //var arg = attr.ConstructorArguments.Single();
                //Assert.Equal(33, arg.Value);
            });
            verifier.VerifyDiagnostics();
        }
```

Second, the type parameters are properly substituted only at the top level. If the references are nested, the type symbols that this function returns are incorrect. This can be observed in the following test:
```
        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66370")]
        public void Attribute_GenericParameter_05()
        {
            var source = """
                public class A2<T> : System.Attribute
                {
                    public A2(B<T>.E t) { }
                }

                public struct B<T>
                {
                    public enum E { }
                    public const E C = (E)33;
                }

                [A2<int>(B<int>.C)]
                class C { }
                """;

            var verifier = CompileAndVerify(source, symbolValidator: static module =>
            {
                var c = module.GlobalNamespace.GetTypeMember("C");
                var attr = c.GetAttributes().Single(d => d.AttributeClass?.Name == "A2");
                Assert.False(attr.HasErrors);
                Assert.Empty(attr.NamedArguments);
                var arg = attr.ConstructorArguments.Single();
                Assert.Equal(33, arg.Value);
                Assert.Equal("B<T>.E", arg.Type.ToTestDisplayString()); // <--- Expected "B<int>.E"
            });
            verifier.VerifyDiagnostics();
        }
```

It looks to me that this helper (`DecodeCustomAttributeParameterTypeOrThrow`) is used exclusively by `MetadataDecoder.GetCustomAttribute` method, which is used in two places:
- `PEAttributeData.EnsureAttributeArgumentsAreLoaded`
- `PENamedTypeSymbol.DecodeAttributeUsageInfo`

In `PEAttributeData.EnsureAttributeArgumentsAreLoaded`, with this change we request `AttributeClass` before calling `MetadataDecoder.GetCustomAttribute`. This also locates a symbol for the constructor. Therefore, we are probably no longer getting much advantage from manually cracking constructor's signature and could simply request types of parameters from the symbol instead.

In `PENamedTypeSymbol.DecodeAttributeUsageInfo`, we are dealing with a well-known attribute and shouldn't be even using `MetadataDecoder.GetCustomAttribute` to crack its application. We have several examples that handle well-known attributes with named arguments. They can be located by looking at the call sites for `PEModule.DecodeCustomAttributeNamedArgumentOrThrow`.

I would do the following:
1. Stop cracking constructor's signature manually for the purpose of `MetadataDecoder.GetCustomAttribute` and pull type information from the constructor symbol itself.
2. Stop calling `MetadataDecoder.GetCustomAttribute` in `PENamedTypeSymbol.DecodeAttributeUsageInfo`, instead have `PEModule` to do all the decoding as it is done for other similar well-known attributes.

Reply:
Thanks! Implemented your suggestion 1. Suggestion 2 could be implemented independently of this PR, I think, plus it seems slightly worse to hard-code the decoding instead of using the general (almost because it cannot handle generic parameters) method.